Split overview/workflow.rst into multiple separate files (#1076)
SUSE/kiwi: doc/source/working_with_kiwi/shell_scripts.rst

type is created from the same unpacked image root tree.

If :file:`images.sh` exits with a non-zero exit code, then KIWI will report
an error and abort the image creation.

See below a common template for :file:`images.sh` script:

.. code-block:: bash

   #======================================
   # Include functions & variables
   #--------------------------------------
   test -f /.kconfig && . /.kconfig
   test -f /.profile && . /.profile

   #======================================
   # Greeting...
   #--------------------------------------
   echo "Configure image: [$kiwi_iname]..."

   #======================================
   # Call configuration code/functions
   #--------------------------------------
   ...

   #======================================
   # Exit successfully
   #--------------------------------------
   exit 0

Functions and variables provided by KIWI
----------------------------------------

KIWI creates the :file:`.kconfig` and :file:`.profile` to be sourced by
:file:`config.sh` and :file:`images.sh`. The former contains various helper
functions which can be used to simplify the image configuration and the

> latter contains environment variables which get populated from the settings

Author: I was referring to the two files `.kconfig` and `.profile`. I've added their names directly into the text.

  in the bin directories.

``baseUpdateSysConfig {filename} {variable} {value}``
  Update the contents of a sysconfig variable

``suseConfig``
  This function is deprecated and is a NOP.

``baseSystemdServiceInstalled {service}``
  Prints the path of the first found systemd unit or mount with name passed
  as the first parameter.

``baseSysVServiceInstalled {service}``
  Prints the name `${service}` if a SysV init service with that name is
  found, otherwise it prints nothing.

``baseSystemdCall {service_name} {args}``
  Calls `systemctl ${args} ${service_name}` if a systemd unit, a systemd
  mount or a SysV init service with the `${service_name}` exist.

``baseInsertService {servicename}``
  Activate the given service via :command:`systemctl`.

``baseRemoveService {servicename}``
  Deactivate the given service via :command:`systemctl`.

``baseService {servicename} {on|off}``
  Activate or deactivate a service via :command:`systemctl`.
  The function requires the service name and the value ``on`` or ``off`` as
  parameters.

  Example to enable the sshd service on boot:

  .. code-block:: bash

     baseService sshd on

``suseInsertService {servicename}``

>   This function calls baseInsertService and exists only for

Reviewer: ```suggestion
  Calls baseInsertService and exists only for
```

  compatibility reasons.

``suseRemoveService {servicename}``

>   This function calls baseRemoveService and exists only for

Reviewer: ```suggestion
  Calls baseRemoveService and exists only for
```

  compatibility reasons.

``suseService {servicename} {on|off}``

>   This function calls baseService and exists only for compatibility

Reviewer: ```suggestion
  Calls baseService and exists only for compatibility
```

  reasons.

``suseActivateDefaultServices``
  Activates the `network` and `cron` services to run at boot.

``suseSetupProduct``

>   This function creates the :file:`/etc/products.d/baseproduct` link

Reviewer: ```suggestion
  Creates the :file:`/etc/products.d/baseproduct` link
```

  pointing to the product referenced by either :file:`/etc/SuSE-brand` or
  :file:`/etc/os-release` or the latest `.prod` file available in
  :file:`/etc/products.d`

``suseSetupProductInformation``

>   This function uses zypper to search for the installed product

Reviewer: ```suggestion
  Uses :command:`zypper` to search for the installed product
```

  and installs all product specific packages. This function fails
  when :command:`zypper` is not the appliances package manager.

``Debug {message}``
  Helper function to print the supplied message if the variable DEBUG is
  set to 1.

``Echo {echo commandline}``
  Helper function to print a message to the controlling terminal.

``Rm {list of files}``
  Helper function to delete files and log the deletion.

``Rpm {rpm commandline}``
  Helper function for calling ``rpm``: forwards all commandline arguments to
  ``rpm`` and logs the call.

> Functions for custom non-dracut based boot

Reviewer: ```suggestion
Functions for Custom non-dracut Based Boot
```

''''''''''''''''''''''''''''''''''''''''''

KIWI also provides the following functions (mostly for compatibility
reasons) which can be used to customize the boot process when using the
custom boot option (see
:ref:`working-with-kiwi-customizing-the-boot-process`):

``baseStripInitrd``

>   This function removes various tools binaries and libraries which

Reviewer: ```suggestion
  Removes various tools binaries and libraries which
```

  are not required to boot a SUSE system with KIWI. This function is not
  required when using the dracut initrd system and is kept for
  compatibility reasons.

``baseStripFirmware``

>   check all kernel modules if they require a firmware and strip out all

Reviewer: ```suggestion
  Check all kernel modules if they require a firmware and strip out all
```

  firmware files which are not referenced by a kernel module

``baseStripModules``

>   search for updated modules and remove the old version which might be

Reviewer: ```suggestion
  Search for updated modules and remove the old version which might be
```

  provided by the standard kernel

``baseStripKernel``

>   this function strips the kernel

Reviewer: ```suggestion
  Strips the kernel:
```

  1. create the :file:`vmlinux.gz` and :file:`vmlinuz` files which are used
     as a fallback for the kernel extraction

  2. handle `<strip type="delete">` requests. Because this information is
     generic not only files of the kernel are affected but also other data
     which are unwanted get deleted here

  3. only keep kernel modules matching the `<drivers>` patterns from the
     kiwi boot image description

  4. lookup kernel module dependencies and bring back modules which were
     removed but still required by other modules that were kept in the
     system

  5. search for duplicate kernel modules due to kernel module updates and
     keep only the latest version

  6. search for kernel firmware files and keep only those for which a
     kernel driver is still present in the system

``suseStripKernel``

>   This function removes all kernel drivers which are not listed in the

Reviewer: ```suggestion
  Removes all kernel drivers which are not listed in the
```

from a given unpacked image (:file:`config.sh` runs prior to the *create
step*). The script should add operating system configuration files which
would be otherwise added by a user driven installer, like the activation of
services, creation of configuration files, preparation of an environment
for a firstboot workflow, etc.

The :file:`config.sh` script is called at the end of the :ref:`prepare step
<prepare-step>` (after users have been set and the *overlay tree directory*
has been applied). If :file:`config.sh` exits with a non-zero exit code
then KIWI will report the failure and abort the image creation.

Find a common template for `config.sh` script below:

.. code:: bash

   #======================================
   # Include functions & variables
   #--------------------------------------
   test -f /.kconfig && . /.kconfig
   test -f /.profile && . /.profile

   #======================================
   # Greeting...
   #--------------------------------------
   echo "Configure image: [$kiwi_iname]..."

   #======================================
   # Mount system filesystems
   #--------------------------------------

>    baseMount

Reviewer 2: @schaefi I beliebe this `baseMount` and `baseCleanMount` are redundant here. I know we have them in many of our example. But I think it might be the time of stop using them in our examples. Current kiwi already calls config.sh in a nice chroot that is safely cleaned afterwards. I believe the less magic we add in `config.sh` the better.

Author: If these functions are redundant, then I'd be more than happy to remove them from the examples.

Reviewer 3: Yes they are redundant. At least the filesystems covered by baseMount are handled by kiwi itself since a long time. If we want to cleanup existing examples in this regard we should however do this in a follow up PR

Reviewer 3: I've update that part of the docs

   #======================================
   # Call configuration code/functions
   #--------------------------------------
   ...

   #======================================
   # Umount kernel filesystems
   #--------------------------------------
   baseCleanMount

   #======================================
   # Exit successfully
   #--------------------------------------
   exit 0

Configuration Tips
^^^^^^^^^^^^^^^^^^

> #. **Stateless systemd UUIDs:**

Reviewer 2: I wrote that part quite a while ago... I am wondering if it is still the case. @dcermak just in case, have it faced or tested it? I'll try to verify this is still valid after going through this PR.

Author: No I haven't faced it at all, I just took that from the existing docs. Not sure if I even rephrased this, so it might very well be your original text.
